### ISSUE TITLE
Smoke and CO?

### ISSUE THREAD

**Issue by dcmeglio** [nutechsoftware/alarmdecoder-smartthings#33]
I know there is a generic smoke detector that triggers if the alarm detects fire. However, if I setup a zone for my smoke and CO zone #s, even though it'll show up as a contact sensor, will it properly report smoke/CO?

**Comment by f34rdotcom** [comment]
Are you referring to Smart Home Monitor as the reporting system?

**Comment by dcmeglio** [comment]
Pretty much, except Hubitat not SmartThings. Basically I want my Hubitat to send me an alert like "Fire detected Upstairs" but how do I know which smoke detector detected smoke? My thought was if I made the AlarmDecoder Service look at my 3 zones for smoke detectors, then the contact would "open" if there is a fire so I could detect which zone trigger the smoke/CO detection. Would that actually work?

**Comment by f34rdotcom** [comment]
I am testing this idea. It's a good idea if it works. May take some tweaks
but it would solve a lot of problems. Not just SHM or HSM some things only
as you noted look at specific device types.


Best
Sean M

**Comment by dcmeglio** [comment]
That'd be cool. I'm fine hacking up the code to make it a different device type (I already did that for my motion detectors) -- what I wasn't sure was, would the app receive what essentially amounts to a zone open/close if a fire/CO is detected? I can add some sendEvent logic to make it send a smoke event instead of a contact event, just don't know if the Ademco would send smoke/CO in a way the app would detect.

**Comment by f34rdotcom** [comment]
I need to brain on this so please help me find a good method. I like the
idea of being able to just change the code for the device in the graph
page. Seems easy and AFAIK would also work on Hubitat? The trick will be
that each device type needs a different type of event sent. As you noted
clear/smoke? is sent to this device type. As apposed to open/close for
contacts.

If it were able to know what type of device exists at that address then we
could modify the message to match. That would allow easy path for changing
one of the virtual zone devices to what is most appropriate.

Best
 SM

**Comment by dcmeglio** [comment]
I like that idea! What do you think of this... In hubitat and SmartThings the device object has a .hasCapability(). So we can check does it have switch does it have smoke, etc. and then we can trigger the correct event.

https://docs.hubitat.com/index.php?title=Device_Object
https://docs.smartthings.com/en/latest/ref-docs/device-ref.html#hascapability

What do you think?

**Comment by f34rdotcom** [comment]
This sounds perfect! I am finishing up a big update to the webapp and then will cycle back into ST/HT coded. I have a good list if improvements to dig into now.

**Comment by f34rdotcom** [comment]
Ok easy button I am using the new RFX processing in ST to create a virtual device and testing change device type and tinkering on code.

**Comment by f34rdotcom** [comment]
Just pushed into dev a proof of concept to be tested on HT. Please let me know if you need me to apply that to the HT github repo. I hope to avoid that till this is tested :) AFAIK the code is a directly drop in just change settings.

**Comment by dcmeglio** [comment]
So I installed the latest code but I don't really know how this RFX stuff works... it's different than the built in 20 zones that get monitored?

**Comment by f34rdotcom** [comment]
Ya sorry it will only work with that commit if you have a 5800 RF
transmitter on your panel and sensors.

I am almost done applying this to all contacts. Hopeful just a few hours
more testing.

**Comment by dcmeglio** [comment]
Great, I'd be happy to test when it's ready

**Comment by f34rdotcom** [comment]
Ok did a bunch of testing and it "seems" to be solid. Setting off fire, alarms, arming disarming while I change device types to see if it does as expected. I am pre sending some events such as "Smoke" translated to On such that for a contact it would e "Open" when On by default. Inverting it as needed for specific applications is now just a toggle of the invert option.

**Comment by dcmeglio** [comment]
It didn't work with Hubitat because of the invert which was trying to access settings. Hubitat doesn't support that. I also came up with a way to set the HSM/SHM automatically. You still have to comment the physicalgraph stuff, but one less manual step now... Also, I added virtual motion, shock, and CO detectors using the new functionality you added. This let me add my motion, glass break, and carbon monoxide detectors too.
